Sources/Customization/DefaultEventProcessor.swift

```diff
@@ -0,0 +1,271 @@
+    var backingStore: DataStoreType
+    var backingStoreName: String
+    
+    // for dispatching events
+    let dispatcher = DispatchQueue(label: "DefaultEventDispatcherQueue")
+    // using a datastore queue with a backing file
+    let dataStore: DataStoreQueueStackImpl<EventForDispatch>
+    // timer as a atomic property.
+    var timer: AtomicProperty<Timer> = AtomicProperty<Timer>()
+    
+    var observerProjectId: NSObjectProtocol?
+    var observerRevision: NSObjectProtocol?
+    
+    public init(eventHandler: OPTEventHandler = DefaultEventHandler(),
+                batchSize: Int = DefaultValues.batchSize,
+                timerInterval: TimeInterval = DefaultValues.timeInterval,
+                maxQueueSize: Int = DefaultValues.maxQueueSize,
+                backingStore: DataStoreType = .file,
+                dataStoreName: String = "OPTEventQueue") {
+        self.eventHandler = eventHandler
+        self.batchSize = batchSize > 0 ? batchSize : DefaultValues.batchSize
+        self.timerInterval = timerInterval >= 0 ? timerInterval : DefaultValues.timeInterval
+        self.maxQueueSize = maxQueueSize >= 100 ? maxQueueSize : DefaultValues.maxQueueSize
+        
+        self.backingStore = backingStore
+        self.backingStoreName = dataStoreName
+
+        switch backingStore {
+        case .file:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreFile<[Data]>(storeName: backingStoreName))
+        case .memory:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreMemory<[Data]>(storeName: backingStoreName))
+        case .userDefaults:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreUserDefaults())
+        }
+        
+        if self.maxQueueSize < self.batchSize {
+            self.logger.e(.eventDispatcherConfigError("batchSize cannot be bigger than maxQueueSize"))
+            self.maxQueueSize = self.batchSize
+        }
+        
+        addProjectChangeNotificationObservers()
+        
+        subscribe()
+    }
+    
+    deinit {
+        stopTimer()
+        
+        removeProjectChangeNotificationObservers()
+
+        unsubscribe()
+    }
+    
+    open func process(event: BatchEvent, completionHandler: DispatchCompletionHandler? = nil) {
+        guard dataStore.count < maxQueueSize else {
+            let error = OptimizelyError.eventDispatchFailed("EventQueue is full")
+            self.logger.e(error)
+            completionHandler?(.failure(error))
+            return
+        }
+        
+        guard let body = try? JSONEncoder().encode(event) else {
```

Comment:
this is now on the main thread if you are calling process from event processor instead of handler.

Reply:
It's called asynchronously on background thread from OptimizelyClient. Check out OptimizelyClient.